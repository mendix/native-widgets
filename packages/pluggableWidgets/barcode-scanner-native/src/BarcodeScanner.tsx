import { flattenStyles } from "@mendix/piw-native-utils-internal";
import { ValueStatus } from "mendix";
import { createElement, ReactElement, useMemo } from "react";
import { View } from "react-native";
import { Camera, useCodeScanner, Code, useCameraDevice } from "react-native-vision-camera";
import BarcodeMask from "react-native-barcode-mask";

import { BarcodeScannerProps } from "../typings/BarcodeScannerProps";
import { BarcodeScannerStyle, defaultBarcodeScannerStyle } from "./ui/styles";
import { executeAction } from "@mendix/piw-utils-internal";

export type Props = BarcodeScannerProps<BarcodeScannerStyle>;

export function BarcodeScanner(props: Props): ReactElement {
    const device = useCameraDevice("back");

    const styles = useMemo(() => flattenStyles(defaultBarcodeScannerStyle, props.style), [props.style]);

    const codeScanner = useCodeScanner({
        codeTypes: ["ean-13", "qr", "aztec", "codabar", "code-128", "data-matrix"],
        onCodeScanned: (codes: Code[]) => {
            if (props.barcode.status !== ValueStatus.Available || codes.length === 0 || !codes[0].value) {
                return;
            }
            const { value } = codes[0];

            if (value !== props.barcode.value) {
                props.barcode.setValue(value);
            }
            executeAction(props.onDetect);
        }
    });

    return (
        <View style={styles.container}>
            {device && (
                <Camera
                    testID={props.name}
                    style={{ flex: 1, justifyContent: "center", alignItems: "center" }}
                    audio={false}
                    isActive
                    device={device}
                    codeScanner={codeScanner}
                >
                    {props.showMask && (
                        <BarcodeMask
                            edgeColor={styles.mask.color}
                            width={styles.mask.width}
                            height={styles.mask.height}
                            backgroundColor={styles.mask.backgroundColor}
                            showAnimatedLine={props.showAnimatedLine}
                        />
                    )}
                </Camera>
            )}
        </View>
    );
}
