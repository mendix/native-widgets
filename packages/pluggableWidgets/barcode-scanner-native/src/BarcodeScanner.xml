<?xml version="1.0" encoding="utf-8"?>
<widget id="com.mendix.widget.native.barcodescanner.BarcodeScanner" supportedPlatform="Native" needsEntityContext="true" offlineCapable="true" pluginWidget="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd">
    <name>Barcode scanner</name>
    <description>Scan barcode and QR code values.</description>
    <studioProCategory>Device features</studioProCategory>
    <studioCategory>Device features</studioCategory>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Data source">
                <property key="barcode" type="attribute">
                    <caption>Barcode</caption>
                    <description>The attribute that will receive the scanned barcode value.</description>
                    <attributeTypes>
                        <attributeType name="String"/>
                    </attributeTypes>
                </property>
            </propertyGroup>
            <propertyGroup caption="Mask">
                <property key="showMask" type="boolean" required="true" defaultValue="false">
                    <caption>Show barcode mask</caption>
                    <description />
                </property>
                <property key="showAnimatedLine" type="boolean" required="true" defaultValue="false">
                    <caption>Show animated line</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="Camera">
                <property key="cameraType" type="enumeration" required="true" defaultValue="back">
                    <caption>Camera type</caption>
                    <description>Select camera to use for the scanner.</description>
                    <enumerationValues>
                        <enumerationValue key="front">Front</enumerationValue>
                        <enumerationValue key="back">Back</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>
            <propertyGroup caption="Events">
                <property key="onDetect" type="action" required="false">
                    <caption>On detect</caption>
                    <description/>
                </property>
            </propertyGroup>
            <propertyGroup caption="Common">
                <systemProperty key="Name"/>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
