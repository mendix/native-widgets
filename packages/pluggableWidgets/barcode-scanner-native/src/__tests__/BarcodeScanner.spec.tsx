// __tests__/BarcodeScanner.spec.tsx
import { actionValue, EditableValueBuilder } from "@mendix/piw-utils-internal";
import { createElement } from "react";
import { View } from "react-native";
import { fireEvent, render, RenderAPI } from "@testing-library/react-native";
import { BarcodeScanner, Props } from "../BarcodeScanner";
import { Camera } from "react-native-vision-camera";

jest.mock("react-native-vision-camera", () => ({
    Camera: ({ children, ...props }: any) => <View {...props}>{children}</View>,
    useCameraDevice: () => "mock-device",
    useCodeScanner: () => jest.fn()
}));

jest.mock("react-native-barcode-mask", () => "BarcodeMask");

describe("BarcodeScanner", () => {
    let defaultProps: Props;

    beforeEach(() => {
        jest.useFakeTimers();
        defaultProps = {
            showAnimatedLine: false,
            showMask: false,
            name: "barcode-scanner-test",
            style: [],
            barcode: new EditableValueBuilder<string>().build()
        };
    });

    afterEach(() => {
        jest.clearAllTimers();
    });

    it("renders", () => {
        const component = render(<BarcodeScanner {...defaultProps} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders with mask", () => {
        const component = render(<BarcodeScanner {...defaultProps} showMask />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders with mask with animated line", () => {
        const component = render(<BarcodeScanner {...defaultProps} showMask showAnimatedLine />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("sets a value and executes the onDetect action when a new barcode is scanned", () => {
        const onDetectAction = actionValue();
        const component = render(<BarcodeScanner {...defaultProps} onDetect={onDetectAction} />);

        detectBarcode(component, "value");
        jest.advanceTimersByTime(2000);

        expect(defaultProps.barcode.setValue).toHaveBeenCalledWith("value");
        expect(onDetectAction.execute).toHaveBeenCalledTimes(1);

        detectBarcode(component, "value1");
        jest.advanceTimersByTime(2000);

        expect(defaultProps.barcode.setValue).toHaveBeenCalledWith("value1");
        expect(onDetectAction.execute).toHaveBeenCalledTimes(2);
    });
});

function detectBarcode(component: RenderAPI, barcode: string): void {
    fireEvent(component.UNSAFE_getByType(Camera), "onCodeScanned", [
        {
            value: barcode,
            type: "qr"
        }
    ]);
}
