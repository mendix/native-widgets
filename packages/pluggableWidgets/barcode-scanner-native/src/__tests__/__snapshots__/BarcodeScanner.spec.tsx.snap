// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarcodeScanner renders 1`] = `
<View
  style={
    {
      "flex": 1,
      "flexDirection": "column",
      "minHeight": 100,
    }
  }
>
  <View
    audio={false}
    codeScanner={[MockFunction]}
    device="mock-device"
    isActive={true}
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
    testID="barcode-scanner-test"
  />
</View>
`;

exports[`BarcodeScanner renders with mask 1`] = `
<View
  style={
    {
      "flex": 1,
      "flexDirection": "column",
      "minHeight": 100,
    }
  }
>
  <View
    audio={false}
    codeScanner={[MockFunction]}
    device="mock-device"
    isActive={true}
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
    testID="barcode-scanner-test"
  >
    <BarcodeMask
      backgroundColor="rgba(0, 0, 0, 0.6)"
      edgeColor="#62B1F6"
      showAnimatedLine={false}
    />
  </View>
</View>
`;

exports[`BarcodeScanner renders with mask with animated line 1`] = `
<View
  style={
    {
      "flex": 1,
      "flexDirection": "column",
      "minHeight": 100,
    }
  }
>
  <View
    audio={false}
    codeScanner={[MockFunction]}
    device="mock-device"
    isActive={true}
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
    testID="barcode-scanner-test"
  >
    <BarcodeMask
      backgroundColor="rgba(0, 0, 0, 0.6)"
      edgeColor="#62B1F6"
      showAnimatedLine={true}
    />
  </View>
</View>
`;
